include/pybind11/stl.h

```diff
@@ -305,6 +316,55 @@ template<> struct type_caster<std::experimental::nullopt_t>
     : public void_caster<std::experimental::nullopt_t> {};
 #endif
 
+#if defined(PYBIND11_HAS_FILESYSTEM)
+template<> struct type_caster<std::filesystem::path> {
+
+    static handle cast(const std::filesystem::path& path, return_value_policy, handle) {
+        PyObject* py_str = [](const auto& w) {
+            using char_type = std::filesystem::path::value_type;
+            if constexpr (std::is_same_v<char_type, char>) {
+                return PyUnicode_DecodeFSDefaultAndSize(w.c_str(), ssize_t(w.size()));
+            } else if constexpr (std::is_same_v<char_type, wchar_t>) {
+                return PyUnicode_FromWideChar(w.c_str(), ssize_t(w.size()));
+            }
+        } (path.native());
+        if (!py_str) {
+            return nullptr;
+        }
+        return module::import("pathlib").attr("Path")(reinterpret_steal<object>(py_str))
+               .release();
+    }
+
+    bool load(handle handle, bool) {
+        using char_type = std::filesystem::path::value_type;
+        PyObject* tmp = nullptr;
+        if constexpr (std::is_same_v<char_type, char>) {
+            if (PyUnicode_FSConverter(handle.ptr(), &tmp)) {
+                if (auto c_str = PyBytes_AsString(tmp)) {
+                    value = c_str;
+                }
+            }
+        } else if constexpr (std::is_same_v<char_type, wchar_t>) {
+            if (PyUnicode_FSDecoder(handle.ptr(), &tmp)) {
+                if (auto c_str = PyUnicode_AsWideCharString(tmp, nullptr)) {
+                    value = c_str;
+                    PyMem_Free(c_str);
+                }
+            }
+        }
+        Py_XDECREF(tmp);
+        if (PyErr_Occurred()) {
+            PyErr_Clear();
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    PYBIND11_TYPE_CASTER(std::filesystem::path, _("PathLike"));
```

Comment:
`_("PathLike")` or `_("os.PathLike")`? This is shown in the signatures, I believe (is it used for anything else?), and I would expect the module there. But maybe that expectation is wrong?

Reply:
os.PathLike seems fine (consistently e.g. with what datetime does).

Reply:
By the way, best would probably be to have os.PathLike for parameter types (as that's what we accept) but pathlib.Path for return types (as that's what we return), but I don't know if that's something there's machinery for right now?

Reply:
I thought about that too; I don't think pybind11 currently allows for this. It should be possible, though, but probably for another PR.

At any rate, there are other places where signatures aren't perfect, when it would cost too much code or performance/space overhead, so it's not a huge worry.